### ISSUE TITLE
"Unimplemented:  Deterministic GPU implementation of unsorted segment reduction op not available" with AUC metric and TF_DETERMINISTIC_OPS

### ISSUE THREAD

**Issue by bersbersbers** [tensorflow/tensorflow#51978]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): OpenSUSE LEAP 15.2
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.6.0-rc2-32-g919f693420e 2.6.0
- Python version: Python 3.9.6
- CUDA/cuDNN version: 11.2 and 8.1.1, I believe
- GPU model and memory: Quadro RTX 6000

Reproduces on Colab with GPU.

**Describe the current behavior**

```
Traceback (most recent call last):
[...]
  File "/home/bers/proj/bug.py", line 12, in <module>
    model.fit(x=data, y=data)
  File "/data2/bers/opt/pyenv/versions/3.9.6/lib/python3.9/site-packages/keras/engine/training.py", line 1184, in fit
    tmp_logs = self.train_function(iterator)
  File "/data2/bers/opt/pyenv/versions/3.9.6/lib/python3.9/site-packages/tensorflow/python/eager/def_function.py", line 885, in __call__
    result = self._call(*args, **kwds)
  File "/data2/bers/opt/pyenv/versions/3.9.6/lib/python3.9/site-packages/tensorflow/python/eager/def_function.py", line 950, in _call
    return self._stateless_fn(*args, **kwds)
  File "/data2/bers/opt/pyenv/versions/3.9.6/lib/python3.9/site-packages/tensorflow/python/eager/function.py", line 3039, in __call__
    return graph_function._call_flat(
  File "/data2/bers/opt/pyenv/versions/3.9.6/lib/python3.9/site-packages/tensorflow/python/eager/function.py", line 1963, in _call_flat
    return self._build_call_outputs(self._inference_function.call(
  File "/data2/bers/opt/pyenv/versions/3.9.6/lib/python3.9/site-packages/tensorflow/python/eager/function.py", line 591, in call
    outputs = execute.execute(
  File "/data2/bers/opt/pyenv/versions/3.9.6/lib/python3.9/site-packages/tensorflow/python/eager/execute.py", line 59, in quick_execute
    tensors = pywrap_tfe.TFE_Py_Execute(ctx._handle, device_name, op_name,
tensorflow.python.framework.errors_impl.UnimplementedError: 2 root error(s) found.
  (0) Unimplemented:  Deterministic GPU implementation of unsorted segment reduction op not available.
	 [[node UnsortedSegmentSum (defined at home/bers/proj/bug.py:12) ]]
	 [[assert_less_equal/Assert/AssertGuard/pivot_f/_13/_39]]
  (1) Unimplemented:  Deterministic GPU implementation of unsorted segment reduction op not available.
	 [[node UnsortedSegmentSum (defined at home/bers/proj/bug.py:12) ]]
0 successful operations.
0 derived errors ignored. [Op:__inference_train_function_513]

Function call stack:
train_function -> train_function
```

**Describe the expected behavior**

No error (works in TF 2.5.0)

**Standalone code to reproduce the issue**
```python
import os

os.environ["TF_DETERMINISTIC_OPS"] = "True"

import tensorflow as tf

data = tf.ones((1, 1))
layer = tf.keras.layers.Input(shape=[1])
model = tf.keras.models.Model(inputs=layer, outputs=layer)
model.compile(loss="categorical_crossentropy", metrics="AUC")
model.fit(x=data, y=data)
```

**Comment by mohantym** [comment]
Hi @bersbersbers ! I could not  replicate the issue in Stable version [2.5](https://colab.research.google.com/gist/mohantym/2269d888b8eb38da4a564023327fa800/github_51978.ipynb#scrollTo=fTqv7A-IlDek) ,[2.6](https://colab.research.google.com/gist/mohantym/7e3a1d7e67fa1d1cfc193afe779710d7/github_51978.ipynb) and 2.6rc2.0 and [nightly ](https://colab.research.google.com/gist/mohantym/5120ec9094a96646004e02ed134b908e/github_51978.ipynb#scrollTo=Lsq-8jz6lXAI)though. Can you please post this issue on [keras-team/keras repo.](https://github.com/keras-team/keras/issues)
To know more see;
[https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999](https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999) . Thanks!

**Comment by bersbersbers** [comment]
@mohamedadaly as I said, you need to test using GPU:
Edit - Notebook settings - Hardware accelerator - GPU, Save.

Then, you can repro in Colab:
![image](https://user-images.githubusercontent.com/12128514/133073490-f8548fa1-cda3-4fdd-bdba-017e5d54b7e9.png)


Can you transfer this issue to keras-team/keras?

**Comment by edwardyehuang** [comment]
Same issue in https://github.com/tensorflow/tensorflow/issues/51466

**Comment by bersbersbers** [comment]
@edwardyehuang correct. That bug has no repro steps, however, so my guess is that that issue should be merged into this issue.

**Comment by mohantym** [comment]
ok @bersbersbers , Is the issue still replicating ? Feel to free to close this issue if it helped.

**Comment by edwardyehuang** [comment]
> 
> 
> ok @bersbersbers , Is the issue still replicating ? Feel to free to close this issue if it helped.

Issue 51466 also did not have solution.

Since TensorFlow 2.5, when TF_DETERMINISTIC_OPS  = 1, calling any of non-deterministic ops (if there is no deterministic implementation ) will raise an "UnimplementedError", this is the root cause.

I think TensorFlow should allow user to use deterministic and non-deterministic togther, before all ops has a deterministic implementation.

**Comment by edwardyehuang** [comment]
@pkanwar23 Will you please look into the issue?

**Comment by bersbersbers** [comment]
@mohantym 
> ok @bersbersbers , Is the issue still replicating ?

Yes it is. It was working in TF2.5 and fails in TF2.6.

**Comment by mohantym** [comment]
Ok! @bersbersbers , Could you see the comment at this [issue1](https://github.com/tensorflow/tensorflow/issues/39751)  ,[issue2](https://stackoverflow.com/questions/50744565/how-to-handle-non-determinism-when-training-on-a-gpu/50746919#50746919) and try again after editing the code like below.


```
import os
os.environ["TF_DETERMINISTIC_OPS"] = "True"

import tensorflow as tf
seed==1441# any random number
tf.random.set_seed(seed)
data = tf.ones((1, 1))
layer = tf.keras.layers.Input(shape=[1])
model = tf.keras.models.Model(inputs=layer, outputs=layer)
model.compile(loss="categorical_crossentropy", metrics="AUC")
model.fit(x=data, y=data)

```

**Comment by bersbersbers** [comment]
@mohantym as you can easily test yourself on Colab, this code raises the same issue (after correcting `seed==1441` to `seed=1441`).

**Comment by edwardyehuang** [comment]
> Ok! @bersbersbers , Could you see the comment at this [issue1](https://github.com/tensorflow/tensorflow/issues/39751) ,[issue2](https://stackoverflow.com/questions/50744565/how-to-handle-non-determinism-when-training-on-a-gpu/50746919#50746919) and try again after editing the code like below.
> 
> ```
> import os
> os.environ["TF_DETERMINISTIC_OPS"] = "True"
> 
> import tensorflow as tf
> seed==1441# any random number
> tf.random.set_seed(seed)
> data = tf.ones((1, 1))
> layer = tf.keras.layers.Input(shape=[1])
> model = tf.keras.models.Model(inputs=layer, outputs=layer)
> model.compile(loss="categorical_crossentropy", metrics="AUC")
> model.fit(x=data, y=data)
> ```

My issue is solved with TF_DISABLE_SEGMENT_REDUCTION_OP_DETERMINISM_EXCEPTIONS = 1

**Comment by mohantym** [comment]
Yes ! @bersbersbers ,Its seed=1441, updated  the above code. Can you try again?

**Comment by bersbersbers** [comment]
@mohantym I already said in https://github.com/tensorflow/tensorflow/issues/51978#issuecomment-918809168 that your corrected code reproduces the issue. There is no point in me trying again.

As also said before, it repros on Colab which *is* using Python 3.7 already. So there is point in trying again with Python 3.8/3.7, either.

**Comment by bersbersbers** [comment]
> My issue is solved with TF_DISABLE_SEGMENT_REDUCTION_OP_DETERMINISM_EXCEPTIONS = 1

This works for me, too:

```python
import os

os.environ["TF_DETERMINISTIC_OPS"] = "True"
os.environ["TF_DISABLE_SEGMENT_REDUCTION_OP_DETERMINISM_EXCEPTIONS"] = "True"

import tensorflow as tf

data = tf.ones((1, 1))
layer = tf.keras.layers.Input(shape=[1])
model = tf.keras.models.Model(inputs=layer, outputs=layer)
model.compile(loss="categorical_crossentropy", metrics="AUC")
model.fit(x=data, y=data)
```

However, I wonder: setting `TF_DISABLE_SEGMENT_REDUCTION_OP_DETERMINISM_EXCEPTIONS` was not necessary in `tensorflow==2.5.1`. So what has changed?
- Was the AUC metric non-deterministic in `tensorflow==2.5.1`, and a missing exception was added in `tensorflow==2.6.0` to make users aware of that fact?
- Was the AUC metric deterministic in `tensorflow==2.5.1`, and this is a regression in `tensorflow==2.6.0`?

**Comment by reedwm** [comment]
This is fixed with https://github.com/tensorflow/tensorflow/pull/51861, and the fix will be in TF 2.7.

I'm unsure if the AUC metric was nondeterministic in TF 2.5. It used `tf.math.unsorted_segment_sum`, which was nondeterminsitic in certain cases, but it's possible AUC did not use it in a nondeterministic way. The exception for `unsorted_segment_sum` was added in TF 2.6, but `unsorted_segment_sum` was nondeterministic before that in certain cases. In any case, this is now fixed, so it's not worth looking into.

**Comment by bersbersbers** [comment]
Verified, thank you all!
